### PR TITLE
Add Dockerfile for Containerizing Preflight

### PR DESCRIPTION
**Pull request by jomkz**
This PR adds the initial Dockerfile to containerize preflight.

## Overview

The preflight binary is built first and the required system packages are then installed, followed by the required external tools (oc, operator-sdk and operator-courier).

In addition, two make targets have been added to build and push the container image. The git hash is used as the image tag, similar to the preflight application version.

Further testing and changes may be required for the podman shell commands to work in all cases but it appears to be working from the limited testing that I have done locally.

## Usage

Running the container without any parameters will display the help.

```
$ podman run -it quay.io/opdev/preflight:f82afd8a7f96c1fcdf1d65a6daed3eeadaba5dbc
A utility that allows you to pre-test your bundles, operators, and container before submitting for Red Hat Certification.
Choose from any of the following checks:
	HasLicense, ValidateOperatorBundle, HasUniqueTag, HasNoProhibitedPackages, RunAsNonRoot, LayerCountAcceptable, HasRequiredLabel, BasedOnUbi, HasMinimalVulnerabilities
Choose from any of the following output formats:
	json, xml, junitxml

Usage:
  preflight <container-image> [flags]

Flags:
  -c, --enabled-checks string   Which checks to apply to the image to ensure compliance.
                                (Env) PREFLIGHT_ENABLED_CHECKS
  -h, --help                    help for preflight
  -o, --output-format string    The format for the check test results.
                                (Env) PREFLIGHT_OUTPUT_FORMAT (Default) json
```

Signed-off-by: John McKenzie <john@mkz.io>

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+# Install dependencies
+RUN microdnf install \
+    buildah \
+    bzip2 \
+    gzip \
+    iptables \
+    podman \
+    python3-pip \
+    skopeo
```
Comment (samira-barouti) [inline Dockerfile:20]: we need `openscap-utils` as well for the vulnerability check

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+# Install OpenShift client binary
+RUN curl -L https://mirror.openshift.com/pub/openshift-v4/clients/ocp/latest/openshift-client-linux-4.7.18.tar.gz | tar -xzv -C /usr/local/bin --exclude=README.md --exclude=kubectl
```
Comment (samira-barouti) [inline Dockerfile:23]: can we ensure to remove the tar file after unzipping?

Comment (jomkz) [reply Dockerfile:23]: The curl call is being piped to tar, so as far as I know the file should not be there after the operation completes. Do you see it in the resulting image somewhere?

File: Dockerfile (hunk 1 above)
Comment (jomkz) [reply Dockerfile:20]: Thanks, I will add it and update the PR.

File: Dockerfile (hunk 2 above)
Comment (skattoju) [reply Dockerfile:25]: i think since its being piped no intermediate tar file gets saved..

File: Dockerfile (hunk 1 above)
Comment (jomkz) [reply Dockerfile:20]: I added the `openscap-utils` dependency, thanks for catching that!

File: Dockerfile (hunk 2 above)
Comment (samira-barouti) [reply Dockerfile:23]: Great! Sorry my OCD kicks in whenever I see `curl` or `wget`

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,35 @@
+## golang image created 2021-06-24T00:31:06.02014601Z 
+FROM docker.io/library/golang@sha256:be99fa59acd78bb22a41bbc1e15ebfab2262498ee0c2e28c3d09bc44d51d1774 AS builder
+WORKDIR /go/src/github.com/komish/preflight
```
Comment (samira-barouti) [inline Dockerfile]: I am just thinking if we move the code to the openshift repo, this needs to be updated. Is there anyway to customize the `WORKDIR`

**Review by samira-barouti — COMMENTED** [review]
One more comment

File: Dockerfile (hunk 3 above)
Comment (jomkz) [reply Dockerfile]: This is an arbitrary directory for building the preflight binary and it does not make it to the final image. It really could be named anything to be honest. My thoughts are that we can just change this when the repo moves.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+# Install OpenShift client binary
+RUN curl -L https://mirror.openshift.com/pub/openshift-v4/clients/ocp/latest/openshift-client-linux-4.7.18.tar.gz | tar -xzv -C /usr/local/bin --exclude=README.md --exclude=kubectl
+
+# Install Operator SDK binray
+RUN curl -Lo /usr/local/bin/operator-sdk https://github.com/operator-framework/operator-sdk/releases/download/v1.9.0/operator-sdk_linux_amd64
+RUN chmod 755 /usr/local/bin/operator-sdk
+
+# Install operator-courier utility
+RUN pip3 install operator-courier==2.1.11
```
Comment (tonytcampbell) [inline Dockerfile]: I'm not sure we need operator-courier.  Someone can correct me but I think we just need operator-sdk, oscap-podman, oc, podman and maybe opm.

Comment (jomkz) [reply Dockerfile]: This brings up a good point. The operator-courier tests are still part of the test playbooks but I don't think courier works with the new bundle format, so I think that is only for operators using the old bundle format. I am unclear if the current process disables the courier test?

File: Dockerfile (hunk 3 above)
Comment (itroyano) [reply Dockerfile]: we can do a build variable  - https://vsupalov.com/docker-env-vars/

value can be taken from Makefile: `$(IMAGE_BUILDER) build -build-arg  WORKDIR=/go/src/github.com/komish/preflight -t $(IMAGE_REPO)/preflight:$(VERSION) . `

however in that case note the behavior seems different with docker build vs. podman build https://github.com/containers/buildah/issues/2888

Comment (jomkz) [reply Dockerfile]: This directory is for the build stage and does not make it into the resulting image or get used in any way, so to me using a variable is adding complexity that is not really needed. The directory could be called `foo` and it really would not matter at the end.  :)  I could just change it to `preflight` or something if this is that big of a deal.

Comment (jomkz) [reply Dockerfile]: I changed the name of the directory used in the build stage to `go/src/preflight`

File: Dockerfile (hunk 4 above)
Comment (jomkz) [reply Dockerfile]: I removed the `operator-courier` dependency for now. I will bring it back in as part of the operator-courier verify check if we determine that is needed after all.

File: Dockerfile (hunk 1 above)
Comment (jomkz) [reply Dockerfile:20]: It does not appear that the `openscap-utils` is available through microdnf in the ubi-minimal base image. However, I see it on my local Fedora 34 machine, so I am wondering if the repo containing that package is not available by default in the ubi-minimal image. I also do not see that package being installed as part of the Vagratfile, unless it is being included as a dependency of something else that we have listed.

Comment (samira-barouti) [reply Dockerfile:20]: Vagrantfile has been updated to install openscap: https://github.com/komish/preflight/pull/37/files
The MR has not been merged yet

**Comment by bcrochet** [comment]
Has been merged into upstream repo. redhat-openshift-ecosystem/openshift-preflight#7

**Pull request closed without merging**